File: backend/app/models/vitals.py
"""
models/vitals.py
------------------
OWNED BY MEMBER 1 (Mubasshira) - Vitals Logging, AI Health Report Analyzer,
AI Symptom Checker, AI Diet Advisor.

VitalsLog keeps its original class/table name and core columns so Member 4's
dashboard queries keep working. Everything below that is new: health_reports
(Feature 2), symptom_logs (Feature 3), and diet_plans/diet_logs (Feature 4 -
a persisted, trend-aware plan with adherence tracking instead of a one-off
Q&A call).
"""

import enum
import uuid
from datetime import datetime

from sqlalchemy import Column, Float, DateTime, ForeignKey, String, LargeBinary, Text, Boolean
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.database import Base


class VitalsLog(Base):
    __tablename__ = "vitals_logs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    blood_pressure = Column(String, nullable=False)   # e.g. "120/80"
    sugar_level = Column(Float, nullable=False)        # mg/dL
    heart_rate = Column(Float, nullable=False)         # bpm
    temperature = Column(Float, nullable=False)        # Celsius
    notes = Column(String, nullable=True)               # optional free-text ("felt dizzy after")
    logged_at = Column(DateTime, default=datetime.utcnow)

    patient = relationship("User", back_populates="vitals")


class HealthReport(Base):
    """Feature 2: AI Health Report Analyzer. Patient uploads a medical PDF,
    we extract its text and ask Gemini to summarize it in plain English."""

    __tablename__ = "health_reports"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    filename = Column(String, nullable=False)
    file_data = Column(LargeBinary, nullable=False)      # original PDF bytes
    extracted_text = Column(Text, nullable=True)          # raw text pulled from the PDF
    ai_summary = Column(Text, nullable=True)               # Gemini's plain-English summary
    created_at = Column(DateTime, default=datetime.utcnow)


class UrgencyLevel(str, enum.Enum):
    normal = "normal"
    monitor = "monitor"
    urgent = "urgent"
    emergency = "emergency"


class SymptomLog(Base):
    """Feature 3: AI Symptom Checker. Unique twist vs. a plain chatbot call -
    the prompt is grounded in the patient's actual recent vitals so advice
    isn't generic, and Gemini is asked to return a structured urgency level.
    An "emergency" result auto-posts to the shared family Notification
    Center (Member 4's feed) instead of silently sitting in a log."""

    __tablename__ = "symptom_logs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    symptoms = Column(Text, nullable=False)
    ai_response = Column(Text, nullable=False)
    urgency = Column(String, nullable=False, default=UrgencyLevel.normal.value)
    escalated = Column(Boolean, default=False)  # True if this triggered a family notification
    created_at = Column(DateTime, default=datetime.utcnow)


class DietPlan(Base):
    """Feature 4: AI Diet Advisor. Unique twist vs. a plain chatbot call -
    the plan is generated from the patient's actual vitals trend (average
    sugar/BP over the last 2 weeks) so it's a genuine recommendation, not a
    generic diet list, and it's persisted so adherence can be tracked
    against it (see DietLog) instead of being a throwaway response."""

    __tablename__ = "diet_plans"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    based_on_summary = Column(Text, nullable=False)   # the vitals-trend context fed to Gemini
    ai_plan = Column(Text, nullable=False)             # structured plan text
    created_at = Column(DateTime, default=datetime.utcnow)

    logs = relationship("DietLog", back_populates="plan", cascade="all, delete-orphan")


class DietLog(Base):
    """Daily adherence check-in against an active DietPlan."""

    __tablename__ = "diet_logs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    plan_id = Column(UUID(as_uuid=True), ForeignKey("diet_plans.id"), nullable=False)
    patient_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    followed = Column(Boolean, nullable=False)
    note = Column(String, nullable=True)
    logged_at = Column(DateTime, default=datetime.utcnow)

    plan = relationship("DietPlan", back_populates="logs")

File: frontend/lib/api/vitals.ts
/**
 * lib/api/vitals.ts
 * --------------------
 * OWNED BY MEMBER 1 (Mubasshira) - Vitals, AI Report Analyzer, Symptom
 * Checker, Diet Advisor.
 */

import { apiFetch, getToken } from "@/lib/apiClient";
import type { VitalsOut } from "./dashboard";

const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export interface VitalsEntryOut extends VitalsOut {
  notes: string | null;
  is_abnormal: boolean;
}

export interface VitalsIn {
  blood_pressure: string;
  sugar_level: number;
  heart_rate: number;
  temperature: number;
  notes?: string;
}

export interface HealthReportOut {
  id: string;
  filename: string;
  ai_summary: string | null;
  created_at: string;
}

export interface SymptomLogOut {
  id: string;
  symptoms: string;
  ai_response: string;
  urgency: "normal" | "monitor" | "urgent" | "emergency";
  escalated: boolean;
  created_at: string;
}

export interface DietPlanOut {
  id: string;
  based_on_summary: string;
  ai_plan: string;
  created_at: string;
}

export interface DietLogOut {
  id: string;
  plan_id: string;
  followed: boolean;
  note: string | null;
  logged_at: string;
}

export interface DietPlanWithLogs {
  plan: DietPlanOut | null;
  logs: DietLogOut[];
  adherence_rate: number | null;
}

// --- Feature 1: Vitals CRUD ---

export function logVitals(payload: VitalsIn) {
  return apiFetch(`/vitals`, { method: "POST", body: JSON.stringify(payload) }) as Promise<VitalsEntryOut>;
}

export function getVitalsHistory(patientId: string, limit = 50) {
  return apiFetch(`/vitals/${patientId}/history?limit=${limit}`) as Promise<VitalsEntryOut[]>;
}

export function updateVitals(vitalsId: string, payload: Partial<VitalsIn>) {
  return apiFetch(`/vitals/${vitalsId}`, { method: "PUT", body: JSON.stringify(payload) }) as Promise<VitalsEntryOut>;
}

export function deleteVitals(vitalsId: string) {
  return apiFetch(`/vitals/${vitalsId}`, { method: "DELETE" });
}

// --- Feature 2: AI Health Report Analyzer ---

export async function uploadHealthReport(file: File): Promise<HealthReportOut> {
  const token = getToken();
  const form = new FormData();
  form.append("file", file);

  const res = await fetch(`${API_BASE}/vitals/reports/upload`, {
    method: "POST",
    headers: token ? { Authorization: `Bearer ${token}` } : {},
    body: form,
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.detail || `Upload failed: ${res.status}`);
  }
  return res.json();
}

export function getHealthReports(patientId: string) {
  return apiFetch(`/vitals/${patientId}/reports`) as Promise<HealthReportOut[]>;
}

// --- Feature 3: AI Symptom Checker ---

export function checkSymptoms(symptoms: string) {
  return apiFetch(`/vitals/symptom-check`, {
    method: "POST",
    body: JSON.stringify({ symptoms }),
  }) as Promise<SymptomLogOut>;
}

export function getSymptomLogs(patientId: string) {
  return apiFetch(`/vitals/${patientId}/symptom-logs`) as Promise<SymptomLogOut[]>;
}

// --- Feature 4: AI Diet Advisor ---

export function generateDietPlan() {
  return apiFetch(`/vitals/diet-plan/generate`, { method: "POST" }) as Promise<DietPlanOut>;
}

export function getLatestDietPlan(patientId: string) {
  return apiFetch(`/vitals/${patientId}/diet-plan/latest`) as Promise<DietPlanWithLogs>;
}

export function logDietAdherence(planId: string, followed: boolean, note?: string) {
  return apiFetch(`/vitals/diet-plan/log`, {
    method: "POST",
    body: JSON.stringify({ plan_id: planId, followed, note }),
  }) as Promise<DietLogOut>;
}
